src/rpc/blockchain.cpp

```diff
@@ -418,6 +420,17 @@ void entryToJSON(UniValue &info, const CTxMemPoolEntry &e)
     }
 
     info.pushKV("spentby", spent);
+
+    // Add opt-in RBF status
+    std::string rbfStatus = "no";
+    RBFTransactionState rbfState = IsRBFOptIn(tx, mempool);
+    if (rbfState == RBF_TRANSACTIONSTATE_UNKNOWN) {
+        rbfStatus = "unknown";
```

Comment:
I think this can just be an `assert(bfState != RBF_TRANSACTIONSTATE_UNKNOWN)`. Looking at [rbf.cpp](https://github.com/bitcoin/bitcoin/blob/master/src/policy/rbf.cpp#L27-L32), the unknown state should only occur if the transaction in question is not found in the mempool, and this function only processes transactions that are currently in the mempool (with mempool lock held). 

If the above is correct (I believe it is), the `bip125-replaceable` field can just become a boolean.

Reply:
I'm not a fan of an assert here. It if it's really not possible, it's better to put the assert in `bf.cpp` and remove this state from the enum.

Reply:
Actually: what about a situation where a descendant is missing from the mempool (because it somehow got evicted)?

Reply:
I should have been more clear. In general it is possible to return `RBF_TRANSACTIONSTATE_UNKNOWN` from `IsRBFOptIn()`, however this call stack in particular (all mempool related rpc's) only calls `IsRBFOptIn()` on transactions we know to be in the mempool. So, the assert would belong here, not in `rfb.cpp`. I understand this still may not be the best option, I just wanted to make my language clear since my comment above was a bit ambiguous. 

I am not sure I understand your situation where 'a descendant is missing from the mempool'. Can you elaborate please?

Reply:
I'm not a fan of using `assert` in non-consensus critical code, given it's massive impact, if it's hit (e.g. by accident in the future etc.).

In the case of the RPCs `"unknown"` isn't triggered, because they iterate over mempool entries, which are inherently part of the mempool, but I kind of tend to leave it there for the potential case `entryToJSON()` is called in a different context. I'd be fine with removing this path though, too.

> Actually: what about a situation where a descendant is missing from the mempool (because it somehow got evicted)?

Not an issue as far as I can see. We only care about ancestors and whether an unconfirmed ancestor signals RBF, which makes the chain replaceable.

Reply:
Agree about the 'descendant is missing from the mempool' not being an issue, or relevant, as far as I can see. 

I also see your point that in general the function could be called in a context where the tx is no longer in the mempool, but in cases strictly calling down from a mempool rpc, it seems like we should always 'know' rbf state in that context. In that sense, it may not make sense for the rpc itself to list `unknown` as an option - unless there actually is a case I am missing where we would not know the rbf state of a tx in our own mempool. 

Having said that, whether or not an `assert` is the right thing to do, I am also an open about. 

I will happily wait for someone more qualified than myself to chime in with their opinion :)

Reply:
> this call stack in particular (all mempool related rpc's) only calls IsRBFOptIn() on transactions we know to be in the mempool

There's no guarantee all ancestors are in the mempool (someone might have deleted mempool.dat and it's still in the process of syncing, unless we guarantee order there). If a transaction in the mempool doesn't signal RBF, but its ancestors do, then it's replaceable. So `IsRBFOptIn` would return UNKNOWN in that case, I think..

Reply:
Mempool.dat is sorted in dependency order, so ancestors always appear earlier. Further, the mempool is at all times internally consistent (when observable): all mempool transactions either spend a transaction output from the chainstate (so from a confirmed transaction), or another mempool output. There are also never any double spends in the mempool.

Reply:
Any opinion on listing `unknown` state @MarcoFalke ?

Reply:
I don't have a strong opinion if the value of 'bip125-replacable' should be string (yes/no/unknown) or bool.

It seems `entryToJSON` can only be called when the entry is in the mempool (c.f. `mempool.vTxHashes[e.vTxHashesIdx].first.ToString()`). Thus, we can surely determine if it is replacable.

Instead of adding an `assert`, I'd prefer to throw some json exception that mentions that it never can be thrown.

I guess you could throw it in `entryToJSON` to make sure it is passed up to all potential callers?

Reply:
But does entryToJSON really care, whether it's in the mempool or not? I think this limits it's functionality in some way.

Reply:
`entryToJSON` specifically processes mempool transactions (`CTxMemPoolEntry &e`) - and I don't believe you should be able to have a CTxMemPoolEntry reference for something not currently in the mempool (it should guarantees this consistency), so in that sense, `entryToJson` should never be called in a "different context" (where the tx is not in mempool). 

Anyhow, I am not trying to get pedantic on this point. I want this change (and requested it) and I am okay with the current implementation.